watchtower-plugin/src/retrier.rs

```diff
@@ -1,171 +1,253 @@
                 }
-                wt_client.set_tower_status(tower_id, crate::TowerStatus::TemporaryUnreachable);
             }
 
-            log::info!("Retrying tower {}", tower_id);
-            match retry_notify(
-                ExponentialBackoff {
-                    max_elapsed_time: Some(Duration::from_secs(self.max_elapsed_time_secs as u64)),
-                    max_interval: Duration::from_secs(self.max_interval_time_secs as u64),
-                    ..ExponentialBackoff::default()
-                },
-                || async { self.add_appointment(tower_id).await },
-                |err, _| {
-                    log::warn!("Retry error happened with {}. {}", tower_id, err);
-                },
-            )
-            .await
-            {
-                Ok(_) => {
-                    log::info!("Retry strategy succeeded for {}", tower_id);
-                    self.wt_client
-                        .lock()
-                        .unwrap()
-                        .set_tower_status(tower_id, crate::TowerStatus::Reachable);
-                }
-                Err(e) => {
-                    log::warn!("Retry strategy gave up for {}. {}", tower_id, e);
-                    // Notice we'll end up here after a permanent error. That is, either after finishing the backoff strategy
-                    // unsuccessfully or by manually raising such an error (like when facing a tower misbehavior)
-                    let mut wt_client = self.wt_client.lock().unwrap();
-                    if let Some(tower) = wt_client.towers.get_mut(&tower_id) {
-                        if tower.status.is_unreachable() {
-                            log::warn!("Setting {} as unreachable", tower_id);
-                            wt_client.set_tower_status(tower_id, crate::TowerStatus::Unreachable);
+            if let Some(retrier) = self.add_pending_appointment(tower_id, locator) {
+                log::info!("Retrying tower {}", tower_id);
+                let wt_client = self.wt_client.clone();
+                let retriers = self.retriers.clone();
+
+                tokio::spawn(async move {
+                    let r = retry_notify(
+                        ExponentialBackoff {
+                            max_elapsed_time: Some(Duration::from_secs(
+                                max_elapsed_time_secs as u64,
+                            )),
+                            max_interval: Duration::from_secs(max_interval_time_secs as u64),
+                            ..ExponentialBackoff::default()
+                        },
+                        || async { retrier.retry_tower(tower_id).await },
+                        |err, _| {
+                            log::warn!("Retry error happened with {}. {}", tower_id, err);
+                        },
+                    )
+                    .await;
+
+                    let mut state = wt_client.lock().unwrap();
+                    retriers.lock().unwrap().remove(&tower_id);
```

Comment:
This can bring a race condition, if the `retry_notify` just finished and also the main retry manager loop got an appointment A from that tower that needs to be retried, the appointment might be added and then the retrier for that tower removed right after. deleting the newly added appointment in the way.

Reply:
Indeed. I've been trying to find a way to fix it but I'm unsure what sync primitive would work here, haven't managed to do so so far :S

Reply:
Not sure if this is the most elegant way, but I think [cfdd3fb](https://github.com/talaia-labs/rust-teos/pull/89/commits/cfdd3fbfba1b2da670c06203e5ca02eb7d522d0b) should have fixed the race condition